pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java

```diff
@@ -45,19 +62,50 @@ public static MetadataStoreExtended createExtended(String metadataURL, MetadataS
+    }
+
+    static void loadProviders() {
+        String factoryClasses = System.getProperty(METADATASTORE_PROVIDERS_PROPERTY);
+        providers = new HashMap<>();
+        providers.put(MEMORY_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(MEMORY_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> new LocalMemoryMetadataStore(url, config)));
+        providers.put(ROCKSDB_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ROCKSDB_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> RocksdbMetadataStore.get(url, config)));
+        providers.put(ETCD_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ETCD_SCHEME_IDENTIFIER,
+                EtcdMetadataStore::new));
+        providers.put(ZK_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ZK_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) ->
+                        new ZKMetadataStore(url.substring(ZK_SCHEME_IDENTIFIER.length()), config, enableWatchers)));
+
+        if (factoryClasses == null) {
+            return;
+        }
+        String[] classNames = factoryClasses.split(",");
+        for (String className : classNames) {
+            try {
+                Class<? extends MetadataStoreProvider> clazz =
+                        (Class<? extends MetadataStoreProvider>) Class.forName(className);
+                MetadataStoreProvider provider = clazz.getConstructor().newInstance();
+                String scheme = provider.urlScheme();
+                providers.put(scheme, provider);
+            } catch (Exception e) {
+                log.warn("Failed to load metadata store provider class for name '{}'", className, e);
+            }
+        }
+    }
 
-        if (metadataURL.startsWith(LocalMemoryMetadataStore.MEMORY_SCHEME_IDENTIFIER)) {
-            return new LocalMemoryMetadataStore(metadataURL, metadataStoreConfig);
-        } else if (metadataURL.startsWith(RocksdbMetadataStore.ROCKSDB_SCHEME_IDENTIFIER)) {
-            return RocksdbMetadataStore.get(metadataURL, metadataStoreConfig);
-        } else if (metadataURL.startsWith(EtcdMetadataStore.ETCD_SCHEME_IDENTIFIER)) {
-            return new EtcdMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
-        } else if (metadataURL.startsWith(ZKMetadataStore.ZK_SCHEME_IDENTIFIER)) {
-            return new ZKMetadataStore(metadataURL.substring(ZKMetadataStore.ZK_SCHEME_IDENTIFIER.length()),
-                    metadataStoreConfig, enableSessionWatcher);
-        } else {
-            return new ZKMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
+    private static MetadataStoreProvider findProvider(String metadataURL) {
+        for (Map.Entry<String, MetadataStoreProvider> entry : providers.entrySet()) {
+            if (metadataURL.startsWith(entry.getKey())) {
+                return entry.getValue();
+            }
         }
+        return null;
```

Comment:
if don't match any provider,  should we guarantee its compatibility to return the zk provider?

https://github.com/apache/pulsar/blob/d4cab6a39209ab26026493855404feefaf9d3a47/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java#L55-L60

```suggestion
        return providers.get(ZK_SCHEME_IDENTIFIER);
```